core/domain/prod_validation_jobs_one_off.py

```diff
@@ -40,13 +42,53 @@ def _external_id_relationships(cls):
             model to validate, and whose values are the corresponding model
             classes.
         """
-        raise NotImplementedError
+        return {}
+
+    @classmethod
+    def _custom_external_id_relationship(cls, item):  # pylint: disable=unused-argument
+        """Define custom external relationships.
+
+        Sometimes we need more complicated logic to determine which external
+        model to fetch. Two example use cases are:
+        1) the field name follows some convention e.g. {exp_id}.{exp_version}
+        2) the field name has to be extracted from a structured field
+
+        Args:
+            item: ndb.Model. Entity to validate.
+
+        Returns:
+            A list of tuples, each consists of a field name (for debug output),
+            the external model class, and a list of keys to fetch. The only
+            requirement of the field names is that they should be unique from
+            the external fields and from each other.
+        """
+        return []
```

Comment:
This should be NotImplementedError (for the base class), to guard against cases where this isn't implemented in a subclass.

Reply:
Would it make sense to raise NotImplementedError for _external_id_relationships (more common) and return a default {} for _custom_external_id_relationships (less common)?

Reply:
Probably less desirable; ideally you want just one pattern for dealing with "missing" methods.

Reply:
How about having both return {}?
If subclass does not implement these, it makes sense that it's implied no external model check is needed

Reply:
I think it's better to force the subclass author to implement it explicitly. That ensures that they at least think about it.